File: common/src/main/java/dev/vini2003/hammer/permission/api/common/manager/RoleManager.java
/*
 * MIT License
 *
 * Copyright (c) 2020 - 2022 vini2003
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package dev.vini2003.hammer.permission.api.common.manager;

import dev.architectury.event.events.common.PlayerEvent;
import dev.architectury.networking.NetworkManager;
import dev.vini2003.hammer.permission.api.common.role.Role;
import dev.vini2003.hammer.permission.registry.common.HPNetworking;
import dev.vini2003.hammer.persistence.api.common.PersistentObject;
import io.netty.buffer.Unpooled;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.network.PacketByteBuf;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.text.TextColor;
import org.jetbrains.annotations.Nullable;

import java.util.*;

public class RoleManager {
	private static class PersistentData extends PersistentObject {
		public List<Role> roles = new ArrayList<>();
		public Map<String, Role> rolesByName = new HashMap<>();
	}
	
	private static final PersistentData PERSISTENT_DATA = PersistentObject.getOrCreate("roles", PersistentData.class);
	
	public static void createRole(Role role) {
		getRoles().add(role);
		getRolesByName().put(role.getName(), role);
		
		PERSISTENT_DATA.save();
	}
	
	public static void createRole(String name, @Nullable String prefix, @Nullable Integer prefixWeight, @Nullable Integer prefixColor) {
		createRole(new Role(name, prefix, prefixWeight, prefixColor));
	}
	
	public static void deleteRole(Role role) {
		getRoles().remove(role);
		getRolesByName().remove(role.getName());
		
		PERSISTENT_DATA.save();
	}
	
	public static Collection<Role> getRoles() {
		return PERSISTENT_DATA.roles;
	}
	
	public static Map<String, Role> getRolesByName() {
		return PERSISTENT_DATA.rolesByName;
	}
	
	public static Role getRoleByName(String name) {
		return getRolesByName().get(name);
	}
	
	public static String getRolePrefix(PlayerEntity player) {
		Role maxRole = null;
		
		for (var role : getRoles()) {
			if (player.hammer$hasRole(role)) {
				if (maxRole == null || role.getPrefixWeight() > maxRole.getPrefixWeight()) {
					maxRole = role;
				}
			}
		}
		
		return maxRole != null ? maxRole.getPrefix() : null;
	}
	
	public static TextColor getRolePrefixColor(PlayerEntity player) {
		Role maxRole = null;
		
		for (var role : getRoles()) {
			if (player.hammer$hasRole(role)) {
				if (maxRole == null || role.getPrefixWeight() > maxRole.getPrefixWeight()) {
					maxRole = role;
				}
			}
		}
		
		return maxRole != null ? TextColor.fromRgb(maxRole.getPrefixColor()) : null;
	}
	
	private static void syncWith(Collection<ServerPlayerEntity> players) {
		for (var role : getRoles()) {
			var buf = new PacketByteBuf(Unpooled.buffer());
			buf.writeString(role.getName());
			
			buf.writeInt(role.getHolders().size());
			
			for (var holder : role.getHolders()) {
				buf.writeUuid(holder);
			}
			
			for (var player : players) {
				NetworkManager.sendToPlayer(player, HPNetworking.SYNC_ROLES, new PacketByteBuf(buf.duplicate()));
			}
		}
		
	}
	
	public static final class JoinListener implements PlayerEvent.PlayerJoin {
		@Override
		public void join(ServerPlayerEntity player) {
			// TODO: Verify.
			var server = player.getServer();
			if (server == null) return;
			
			syncWith(server.hammer$getPlayers());
		}
	}
}

File: hammer-permission/src/main/kotlin/dev/vini2003/hammer/permission/common/manager/RoleManager.kt
package dev.vini2003.hammer.permission.common.manager

import dev.vini2003.hammer.common.util.BufUtils
import dev.vini2003.hammer.permission.common.packet.sync.SyncRolesPacket
import dev.vini2003.hammer.permission.common.role.Role
import dev.vini2003.hammer.permission.common.util.extension.hasRole
import dev.vini2003.hammer.permission.registry.common.HPNetworking
import net.fabricmc.fabric.api.networking.v1.PacketByteBufs
import net.fabricmc.fabric.api.networking.v1.PacketSender
import net.fabricmc.fabric.api.networking.v1.ServerPlayConnectionEvents
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.server.MinecraftServer
import net.minecraft.server.network.ServerPlayNetworkHandler
import net.minecraft.text.TextColor

object RoleManager {
	private val ROLES: MutableList<Role> = mutableListOf()
	private val ROLES_BY_NAME: MutableMap<String, Role> = mutableMapOf()
	
	fun register(role: Role) {
		ROLES.add(role)
		
		ROLES_BY_NAME[role.name] = role
	}
	
	fun getRoleByName(name: String): Role? {
		return ROLES_BY_NAME[name]
	}
	
	@JvmStatic
	fun getRolePrefix(player: PlayerEntity): String? {
		return ROLES.filter { role ->
			player.hasRole(role)
		}.maxByOrNull {
			role -> role.prefixWeight ?: 0
		}?.prefix
	}
	
	@JvmStatic
	fun getRolePrefixColor(player: PlayerEntity): TextColor? {
		return ROLES.filter { role ->
			player.hasRole(role)
		}.maxByOrNull { role ->
			role.prefixWeight ?: 0
		}?.prefixColor?.let { color ->
			TextColor.fromRgb(color)
		}
	}
	
	fun roles(): Iterator<Role> {
		return ROLES.iterator()
	}
	
	object ServerPlayConnectionEventsJoinListener : ServerPlayConnectionEvents.Join {
		override fun onPlayReady(handler: ServerPlayNetworkHandler, sender: PacketSender, server: MinecraftServer) {
			val players = server.playerManager.playerList.toMutableList()
			players.add(handler.player)
			
			val packet = SyncRolesPacket(ROLES_BY_NAME.mapValues { entry -> entry.value.holders.toList() })
			val buf = BufUtils.toPacketByteBuf(packet)
			
			players.forEach { player ->
				ServerPlayNetworking.send(player, HPNetworking.SYNC_ROLES, PacketByteBufs.duplicate(buf))
			}
		}
	}
}



































